### ISSUE TITLE
pyarmor registration is failling on linux. It works on Windows for version 8.4.4

### ISSUE THREAD

**Issue by shahir-imemori** [dashingsoft/pyarmor#1628]
pyarmor gen -O dist -r src/*
INFO     Python 3.10.12
INFO     Pyarmor 8.4.4 (trial), 000000, non-profits
INFO     Platform linux.x86_64
INFO     search inputs ...
INFO     find script src/azure_uuid.py
INFO     find script src/dash.json
INFO     find script src/encrypt.py
INFO     find script src/installer.py
INFO     find script src/main.py
INFO     find script src/password.py
INFO     find package at src/router
INFO     find script src/strings.py
INFO     find 8 top resources
ERROR    invalid license token, try to run `pyarmor reg` to register license again

When running registration again getting
pyarmor reg -p <product name> ./service/pyarmor/pyarmor-regcode-5397.txt 
INFO     Python 3.10.12
INFO     Pyarmor 8.4.4 (trial), 000000, non-profits
INFO     Platform linux.x86_64
INFO     prepare "./service/pyarmor/pyarmor-regcode-5397.txt"
INFO     query key file from server
This license registration information will be

License Type    : pyarmor-pro
License To      : iMemori Technologies Pvt Ltd
License Product : <product name>

This license is about to be used for product <product name>

Are you sure to continue? (yes/no) yes
INFO     process activation file "./service/pyarmor/pyarmor-regcode-5397.txt"
INFO     send request to server
INFO     write registration file "pyarmor-regfile-5397.zip"
INFO     register "pyarmor-regfile-5397.zip"
INFO     update license token
ERROR    invalid license token, try to run `pyarmor reg` to register license again

**Comment by jondy** [comment]
Please do not use `pyarmor-regcode-5397.txt` more than once, use `pyarmor-regcode-5397.zip` instead, `pyarmor-regcode-5397.txt` may not work after initial registration. 

And using debug option `pyarmor -d reg pyarmor-regcode-5397.zip` to get more log, also provide the content of `pyarmor.error.log`

**Comment by shahir-imemori** [comment]
Please find below output
pyarmor -d reg ./service/pyarmor/pyarmor-regfile-5397.zip
INFO     Python 3.10.12
INFO     Pyarmor 8.4.4 (trial), 000000, non-profits
INFO     Platform linux.x86_64
DEBUG    native platform linux.x86_64
DEBUG    home path: /var/lib/jenkins/.pyarmor
INFO     register "./service/pyarmor/pyarmor-regfile-5397.zip"
DEBUG    extracting license.lic
DEBUG    extracting .pyarmor_capsule.zip
INFO     update license token
DEBUG    unknown error, please check pyarmor.error.log
ERROR    invalid license token, try to run `pyarmor reg` to register license again
----------------------- Output of pyarmor.error.log
cat pyarmor.error.log 
1520587 MainProcess 2023-12-10 17:23:51,332
Traceback (most recent call last):
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/__main__.py", line 718, in main
    main_entry(sys.argv[1:])
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/__main__.py", line 706, in main_entry
    return args.func(ctx, args)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/__main__.py", line 305, in cmd_reg
    reg.register_regfile(regfile)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/register.py", line 209, in register_regfile
    self.update_token()
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/register.py", line 116, in update_token
    Pytransform3._update_token(self.ctx)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/core/__init__.py", line 127, in _update_token
    m = Pytransform3.init(ctx)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/core/__init__.py", line 97, in init
    m.init_ctx(ctx)
RuntimeError: invalid license token, try to run `pyarmor reg` to register license again

FYI, The registration using same on my mac works

**Comment by shahir-imemori** [comment]
Please do not close the issue without resolving it

**Comment by jondy** [comment]
Try to `rm -rf ~/.pyarmor` and register it again.

I'll close the issues which missing necessary informations or wrong informations, or the solutions has been documented in the common solutions.

**Comment by shahir-imemori** [comment]
I already tried deleting the folder. Hence raised. The system is having access to the internet access. Below commands demonstrate the same.

:~$ rm -rf ~/.pyarmor
:~$ cd ~/build
$ pyarmor -d reg ./service/pyarmor/pyarmor-regfile-5397.zip
INFO     Python 3.10.12
INFO     Pyarmor 8.4.4 (trial), 000000, non-profits
INFO     Platform linux.x86_64
DEBUG    native platform linux.x86_64
DEBUG    home path: /var/lib/jenkins/.pyarmor
INFO     register "./service/pyarmor/pyarmor-regfile-5397.zip"
DEBUG    extracting license.lic
DEBUG    extracting .pyarmor_capsule.zip
INFO     update license token
DEBUG    unknown error, please check pyarmor.error.log
ERROR    invalid license token, try to run `pyarmor reg` to register license again
jenkins@imem-cs-mgmt-devops:/data/jenkins/workspace/v2-build/build$ 

$ wget https://pyarmor.dashingsoft.com/
--2023-12-11 10:00:09--  https://pyarmor.dashingsoft.com/
Resolving pyarmor.dashingsoft.com (pyarmor.dashingsoft.com)... 119.23.58.77
Connecting to pyarmor.dashingsoft.com (pyarmor.dashingsoft.com)|119.23.58.77|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 12413 (12K) [text/html]
Saving to: ‘index.html’

index.html                                  100%[===========================================================================================>]  12.12K  --.-KB/s    in 0s      

2023-12-11 10:00:10 (463 MB/s) - ‘index.html’ saved [12413/12413]

**Comment by jondy** [comment]
After register failed, what's the output of these commands?
```
ls -lha ~/.pyarmor
ls -lha /var/lib/jenkins/.pyarmor 
```

**Comment by shahir-imemori** [comment]
This is a virtual machine on Azure cloud. We changed the pipeline to build and copy obfuscated code to the docker build so that the docker-related issue doesn't exist.

Below is the requested output

jenkins@<hostname>:~$ ls -alrh ~/.pyarmor
total 24K
-rw-rw-r--  1 jenkins jenkins  197 Dec 12 00:18 license.lic
-rw-rw-r--  1 jenkins jenkins 2.8K Dec 12 00:18 .pyarmor_capsule.zip
-rw-rw-r--  1 jenkins jenkins    0 Dec 12 00:19 .license.token
drwxr-xr-x 34 jenkins jenkins  12K Dec 11 19:41 ..
drwxrwxr-x  2 jenkins jenkins 4.0K Dec 11 09:57 .

**Comment by jondy** [comment]
Try to apply this patch in `/path/to/pyarmor/cli/context`
```patch
--- a/pyarmor/cli/context.py
+++ b/pyarmor/cli/context.py
@@ -203,7 +203,11 @@ class Context(object):
 
     def save_token(self, data):
         with open(self.license_token, 'wb') as f:
-            f.write(data)
+            print('got token: %r' % data)
+            if data.endswith(b'=='):
+                f.write(data)
+            else:
+                n = data.rfind(b'==')
+                f.write(data[:n+2])
 
     def clear_token(self):
         if os.path.exists(self.license_token):
```

**Comment by shahir-imemori** [comment]
Below is the output after patching the code
$ pyarmor -d reg ./service/pyarmor/pyarmor-regfile-5397.zip
INFO     Python 3.10.12
INFO     Pyarmor 8.4.4 (trial), 000000, non-profits
INFO     Platform linux.x86_64
DEBUG    native platform linux.x86_64
DEBUG    home path: /var/lib/jenkins/.pyarmor
INFO     register "./service/pyarmor/pyarmor-regfile-5397.zip"
DEBUG    extracting license.lic
DEBUG    extracting .pyarmor_capsule.zip
INFO     update license token
got token: b'sTkAAAEHAACRd3xlAAAAAHB5YXJtb3ItdmF4LTAwNTM5NwAApZfUzn6enzgZVxuwdkIJZAAAAAAAAAAAAAAAACFiZDI5YmUwNGE3ODliZmZiODUwMTZhNDA4MDJlN2Q5OGYcaU1lbW9yaSBUZWNobm9sb2dpZXMgUHZ0IEx0ZAdpTWVtb3JpAA== MLeax/jHRzgtib/hJi6ynKdVEL7eUG/+SF7YEQo7GFezGcwuVOXKiyjt1pIRAhudKzmNwr34PUSTFoopwHGxF+edp8HCVPoVt3Opp6X2n/OxIk3FuE9K9B/N+UvN6TsX2eqYmGMWt1jLB9HJLWTuBSS0iVpCRSHvM0buVYLv+Rq2ITZocsIIgtWq7gMdVUK0JN5h78Kd8O6mCCLYvJs1CyV4f0G3u5Ci59bRXWKtCmri6YfUd1BBAgpq84Td09qs/j9WcFQsTi/4Ng9MKfFfgMK93Cf+9ufV5iGGigblhivqXx7tMr81JvH7vKkRzdQ1xTNJCJ5MtEteT6YyVNQuUg====\r\n0\r\n\r\n'
DEBUG    unknown error, please check pyarmor.error.log
ERROR    invalid license token, try to run `pyarmor reg` to register license again
-----------------------------------
$ cat pyarmor.error.log 
1630696 MainProcess 2023-12-12 21:28:10,048
Traceback (most recent call last):
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/__main__.py", line 718, in main
    main_entry(sys.argv[1:])
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/__main__.py", line 706, in main_entry
    return args.func(ctx, args)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/__main__.py", line 305, in cmd_reg
    reg.register_regfile(regfile)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/register.py", line 209, in register_regfile
    self.update_token()
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/register.py", line 116, in update_token
    Pytransform3._update_token(self.ctx)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/core/__init__.py", line 127, in _update_token
    m = Pytransform3.init(ctx)
  File "/var/lib/jenkins/.local/lib/python3.10/site-packages/pyarmor/cli/core/__init__.py", line 97, in init
    m.init_ctx(ctx)
RuntimeError: invalid license token, try to run `pyarmor reg` to register license again

I upgraded pyarmor to 8.4.5 and still not working. Please find below details

$ python3 -m pip install pyarmor --upgrade
Defaulting to user installation because normal site-packages is not writeable
Requirement already satisfied: pyarmor in /data/jenkins/python3.10/site-packages (8.4.4)
Collecting pyarmor
  Downloading pyarmor-8.4.5-py3-none-any.whl.metadata (2.5 kB)
Requirement already satisfied: pyarmor.cli.core~=5.4.2 in /data/jenkins/python3.10/site-packages (from pyarmor) (5.4.2)
Downloading pyarmor-8.4.5-py3-none-any.whl (2.7 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.7/2.7 MB 8.5 MB/s eta 0:00:00
Installing collected packages: pyarmor
  Attempting uninstall: pyarmor
    Found existing installation: pyarmor 8.4.4
    Uninstalling pyarmor-8.4.4:
      Successfully uninstalled pyarmor-8.4.4
  WARNING: The scripts pyarmor, pyarmor-7, pyarmor-8 and pyarmor-auth are installed in '/var/lib/jenkins/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed pyarmor-8.4.5

$ pyarmor -d reg ./service/pyarmor/pyarmor-regfile-5397.zip
INFO     Python 3.10.12
INFO     Pyarmor 8.4.5 (trial), 000000, non-profits
INFO     Platform linux.x86_64
DEBUG    native platform linux.x86_64
DEBUG    home path: /var/lib/jenkins/.pyarmor
INFO     register "./service/pyarmor/pyarmor-regfile-5397.zip"
DEBUG    extracting license.lic
DEBUG    extracting .pyarmor_capsule.zip
INFO     update license token
DEBUG    unknown error, please check pyarmor.error.log
ERROR    invalid license token, try to run `pyarmor reg` to register license again

**Comment by jondy** [comment]
Sorry, it will be fixed in next release because it need update core package `pyarmor.cli.core`

Verify it by downloading this pre-release extension

http://pyarmor.dashingsoft.com/downloads/temp/5.4.3/linux.x86_64/libs/cp3.10/pytransform3.so

and overwrite `/path/to/pyarmor/cli/core/pytransform3.so` in runner

**Comment by shahir-imemori** [comment]
Hi Jondy,

Thanks a lot for your support. Pyarmor registration is working now. Is there way we can get the obfuscated source run inside different environment. For example register and generate source inside mac and copy generated content to linux docker.  I am getting error “ImportError: Error loading shared library /app/dist/pyarmor_runtime_005397/pyarmor_runtime.so: Exec format error”. I am doing this to circumvent error which I am having with multiple registration inside each type of docker container.


$ cp /tmp/pytransform3.so ~/.local/lib/python3.10/site-packages/pyarmor/cli/core

$ pyarmor -d reg ./service/pyarmor/pyarmor-regfile-5397.zip
INFO     Python 3.10.12
INFO     Pyarmor 8.4.5 (trial), 000000, non-profits
INFO     Platform linux.x86_64
DEBUG    native platform linux.x86_64
DEBUG    home path: /var/lib/jenkins/.pyarmor
INFO     register "./service/pyarmor/pyarmor-regfile-5397.zip"
DEBUG    extracting license.lic
DEBUG    extracting .pyarmor_capsule.zip
INFO     update license token
INFO     This license registration information:

License Type    : pyarmor-pro
License No.     : pyarmor-vax-005397
License To      : iMemori Technologies Pvt Ltd
License Product : iMemori

BCC Mode        : Yes
RFT Mode        : Yes

Notes
* Need verify license online when obfuscating scripts

------------------------ Following error I am getting inside-----
Traceback (most recent call last):
  File "/app/dist/main.py", line 2, in <module>
    from pyarmor_runtime_005397 import __pyarmor__
  File "/app/dist/pyarmor_runtime_005397/__init__.py", line 2, in <module>
    from .pyarmor_runtime import __pyarmor__
ImportError: Error loading shared library /app/dist/pyarmor_runtime_005397/pyarmor_runtime.so: Exec format error
--
Cheers,
Shahir

From: Jondy ***@***.***>
Date: Wednesday, 13 December 2023 at 7:41 AM
To: dashingsoft/pyarmor ***@***.***>
Cc: Mohamed Shahir ***@***.***>, Author ***@***.***>
Subject: Re: [dashingsoft/pyarmor] pyarmor registration is failling on linux. It works on Windows for version 8.4.4 (Issue #1628)

Sorry, it will be fixed in next release because it need update core package pyarmor.cli.core

Verify it by downloading this pre-release extension

http://pyarmor.dashingsoft.com/downloads/temp/5.4.3/linux.x86_64/libs/cp3.10/pytransform3.so

and overwrite /path/to/pyarmor/cli/core/pytransform3.so in runner

**Comment by jondy** [comment]
Fixed in 8.4.6
